File: detox_project/public/js/stock_entry_manufacture.js
// ABP2-I419 Phase 3 (Sahil 2026-06-17, Out of BRD) — Stock Entry
// Manufacture client logic: L08 process auto-fetch, L09 source-row
// locking, L10 target defaults, L12 PO rate fetch, VAL-09 confirm-on-
// change of process when rows already exist.

const MFG_TYPES = new Set([
    "Manufacture",
    "Material Transfer for Manufacture",
    "Repack",
]);

frappe.ui.form.on("Stock Entry", {
    refresh(frm) {
        _apply_phase3_visibility(frm);
        _populate_process_options(frm);
        _lock_source_row_fields(frm);
        _wire_po_item_filter(frm);
    },

    custom_start_time(frm) { _recompute_production_time(frm); },
    custom_end_time(frm) { _recompute_production_time(frm); },

    stock_entry_type(frm) {
        _apply_phase3_visibility(frm);
    },

    production_plan(frm) {
        _populate_process_options(frm);
    },

    custom_process_selection(frm) {
        if (!_is_mfg(frm)) return;
        if (!frm.doc.production_plan) {
            frappe.msgprint(__("Pick a Production Plan first."));
            return;
        }
        if ((frm.doc.items || []).length > 0) {
            frappe.confirm(
                __("Changing the process will clear all source rows. Continue?"),
                () => _fetch_operation_rows(frm),
            );
        } else {
            _fetch_operation_rows(frm);
        }
    },
});

frappe.ui.form.on("Stock Entry Detail", {
    item_code(frm, cdt, cdn) {
        if (!_is_mfg(frm)) return;
        const row = locals[cdt][cdn];
        // L10 — for target FG rows, pull t_warehouse + basic_rate from
        // the plan's Table 1.
        if (row.is_finished_item && frm.doc.production_plan && row.item_code) {
            frappe.call({
                method: "detox_project.detox_project.overrides.stock_entry_manufacture.get_fg_defaults",
                args: {
                    production_plan: frm.doc.production_plan,
                    item_code: row.item_code,
                },
                callback(r) {
                    const d = (r && r.message) || {};
                    if (d.t_warehouse) {
                        frappe.model.set_value(cdt, cdn, "t_warehouse", d.t_warehouse);
                    }
                    if (d.basic_rate) {
                        frappe.model.set_value(cdt, cdn, "basic_rate", d.basic_rate);
                    }
                },
            });
        }
    },

    custom_purchase_order_item(frm, cdt, cdn) {
        if (!_is_mfg(frm)) return;
        const row = locals[cdt][cdn];
        if (!row.custom_purchase_order_item || !row.custom_purchase_order) return;
        // L12 — fetch the linked PO line's rate so the user can see the
        // actual procured rate next to the plan's standard_rate.
        frappe.db.get_value(
            "Purchase Order Item",
            row.custom_purchase_order_item,
            ["rate", "amount"],
            (r) => {
                if (r && r.rate) {
                    frappe.model.set_value(cdt, cdn, "additional_cost", flt(r.rate));
                }
            },
        );
    },
});

// CR-07 refresh wiring — filter the Additional Costs grid's PO Item picker
// by the row's chosen Purchase Order (CR-07.4). Registered on the Stock
// Entry form refresh via the block below.
frappe.ui.form.on("Stock Entry", {
    refresh(frm) {
        _wire_addl_cost_po_item_filter(frm);
    },
});

function _wire_addl_cost_po_item_filter(frm) {
    const grid = frm.fields_dict.additional_costs && frm.fields_dict.additional_costs.grid;
    if (!grid || !grid.get_field) return;
    const po_item_field = grid.get_field("custom_purchase_order_item");
    if (!po_item_field) return;
    // CR-07.4 — restrict PO Item options to lines of the row's PO. Reuse the
    // same server query the source-row picker uses (shows item_code, not the
    // row hash — Sahil Image #34).
    po_item_field.get_query = function (doc, cdt, cdn) {
        const row = locals[cdt] && locals[cdt][cdn];
        const filters = {};
        if (row && row.custom_purchase_order) {
            filters.parent = row.custom_purchase_order;
        }
        return {
            query: "detox_project.detox_project.overrides.stock_entry_manufacture.po_item_link_query",
            filters: filters,
        };
    };
}

function _is_mfg(frm) {
    return MFG_TYPES.has(frm.doc.stock_entry_type || "");
}

function _apply_phase3_visibility(frm) {
    const on = _is_mfg(frm);
    ["custom_process_selection",
     "custom_production_time",
     "custom_time_uom"].forEach((fn) => {
        if (frm.fields_dict[fn]) frm.toggle_display(fn, on);
        if (frm.fields_dict[fn] && on && fn !== "custom_process_selection") {
            // Make production_time + time_uom mandatory for Manufacturing-flow SEs.
            frm.toggle_reqd(fn, true);
        }
    });
}

// Sahil Image #33 — compute Production Time in minutes from
// (custom_end_time - custom_start_time) and force Time UOM = Minutes.
function _recompute_production_time(frm) {
    const start = frm.doc.custom_start_time;
    const end = frm.doc.custom_end_time;
    if (!start || !end) return;
    const start_ms = frappe.datetime.str_to_obj(start).getTime();
    const end_ms = frappe.datetime.str_to_obj(end).getTime();
    if (!isFinite(start_ms) || !isFinite(end_ms)) return;
    if (end_ms <= start_ms) {
        frappe.msgprint({
            title: __("Invalid range"),
            message: __("End Time must be after Start Time."),
            indicator: "orange",
        });
        return;
    }
    const minutes = (end_ms - start_ms) / 60000;
    frm.set_value("custom_production_time", Math.round(minutes * 100) / 100);
    frm.set_value("custom_time_uom", "Minutes");
}

// FR-11 — when the user picks a Purchase Order on a source row, restrict
// the PO Item picker to lines from THAT PO. Also point at the custom
// query so the autocomplete shows item_code instead of the row's hash
// name (Sahil Image #34).
function _wire_po_item_filter(frm) {
    const grid = frm.fields_dict.items && frm.fields_dict.items.grid;
    if (!grid || !grid.get_field) return;
    const po_item_field = grid.get_field("custom_purchase_order_item");
    if (!po_item_field) return;
    po_item_field.get_query = function (doc, cdt, cdn) {
        const row = locals[cdt] && locals[cdt][cdn];
        const filters = {};
        if (row && row.custom_purchase_order) {
            filters.parent = row.custom_purchase_order;
        }
        return {
            query: "detox_project.detox_project.overrides.stock_entry_manufacture.po_item_link_query",
            filters: filters,
        };
    };
}

// FR-09 — on Manufacture-flow SEs, source rows (rows with s_warehouse
// set) have item_code / uom / basic_rate / expense_account read-only;
// only qty stays editable. Target FG rows keep all fields editable
// for the user to enter what was actually produced.
function _lock_source_row_fields(frm) {
    if (!_is_mfg(frm)) return;
    const grid = frm.fields_dict.items && frm.fields_dict.items.grid;
    if (!grid || !grid.grid_rows) return;
    const LOCKED = ["item_code", "uom", "basic_rate", "expense_account"];
    grid.grid_rows.forEach((gr) => {
        const row = gr.doc;
        if (!row || !row.s_warehouse) return;
        LOCKED.forEach((fn) => {
            try { gr.toggle_editable(fn, false); } catch (e) { /* field absent */ }
        });
    });
}

function _populate_process_options(frm) {
    if (!frm.doc.production_plan) {
        if (frm.fields_dict.custom_process_selection) {
            frm.set_df_property("custom_process_selection", "options", "");
            frm.refresh_field("custom_process_selection");
        }
        return;
    }
    // Whitelisted server method — bypasses client-side child-perm
    // walls on Detox Production Plan Process.
    frappe.call({
        method: "detox_project.detox_project.overrides.stock_entry_manufacture.get_plan_processes",
        args: {production_plan: frm.doc.production_plan},
        callback(r) {
            const rows = (r && r.message) || [];
            const seen = new Set();
            const opts = ["", ...rows.map((row) => row.operation_name).filter((n) => {
                if (!n || seen.has(n)) return false;
                seen.add(n);
                return true;
            })];
            if (frm.fields_dict.custom_process_selection) {
                frm.set_df_property("custom_process_selection", "options", opts.join("\n"));
                frm.refresh_field("custom_process_selection");
            }
        },
    });
}

function _fetch_operation_rows(frm) {
    frappe.call({
        method: "detox_project.detox_project.overrides.stock_entry_manufacture.get_operation_rm_rows",
        args: {
            production_plan: frm.doc.production_plan,
            operation: frm.doc.custom_process_selection,
        },
        callback(r) {
            const rows = (r && r.message) || [];
            // FR-09 — Clear any rows we previously auto-fetched. We tag
            // each fetched row with _auto_fetched so we can wipe them
            // on the next Process change without disturbing target FG
            // rows the user typed manually. Also clear empty placeholder
            // rows (no item_code) Frappe auto-adds to the grid.
            const survivors = (frm.doc.items || []).filter((row) => {
                if (row._auto_fetched) return false;
                if (!row.item_code) return false;  // empty placeholder
                return true;
            });
            // Rebuild the table from scratch — set the canonical array
            // back and let Frappe re-index.
            frm.doc.items = survivors;
            if (!rows.length) {
                frm.refresh_field("items");
                frappe.msgprint({
                    title: __("No materials found"),
                    message: __(
                        "No Materials / Service rows are defined for Operation "
                        + "'{0}' on the linked Production Plan. Add them on the "
                        + "plan first (per-Operation 'Add Material / Service Row' "
                        + "button), then re-pick the Process here.",
                        [frm.doc.custom_process_selection]
                    ),
                    indicator: "orange",
                });
                return;
            }
            rows.forEach((src) => {
                const item = frappe.model.add_child(frm.doc, "Stock Entry Detail", "items");
                item._auto_fetched = true;  // marker for the next clear
                item.item_code = src.item_code;
                item.item_name = src.item_name;
                item.uom = src.uom;
                item.stock_uom = src.stock_uom;
                item.conversion_factor = src.conversion_factor;
                item.basic_rate = src.basic_rate;
                item.qty = 0;
                item.cost_center = src.cost_center;
                item.project = src.project;
                item.custom_budget_category = src.budget_category;
                if (src.expense_account) item.expense_account = src.expense_account;
                if (frm.doc.from_warehouse) item.s_warehouse = frm.doc.from_warehouse;
            });
            frm.refresh_field("items");
            // Lock the fetched source rows per FR-09.
            _lock_source_row_fields(frm);
            frappe.show_alert({
                message: __("Fetched {0} source row(s) for {1}.",
                            [rows.length, frm.doc.custom_process_selection]),
                indicator: "green",
            });
        },
    });
}


// ABP2-I419 reopen item #7 (Raj 2026-06-19) — Additional Cost section
// got Qty + Rate columns. When either changes, recompute Amount = Qty
// × Rate so the user doesn't have to re-do the arithmetic. The Amount
// cell stays editable for cases where qty/rate aren't known and the
// user only has the rolled-up cost — only overwrite Amount when BOTH
// custom_qty and custom_rate are non-zero. The child doctype is
// shared (`Landed Cost Taxes and Charges` is also used by Landed Cost
// Voucher); this handler is gated on the parent type being Stock
// Entry so the LCV form isn't affected.
function _recompute_addl_cost_amount(frm, cdt, cdn) {
    if (frm.doctype !== "Stock Entry") return;
    const row = locals[cdt][cdn];
    const qty = flt(row.custom_qty);
    const rate = flt(row.custom_rate);
    if (qty && rate) {
        frappe.model.set_value(cdt, cdn, "amount", qty * rate);
    }
}

// CR-07.3 — when a Service Item is picked on an additional-cost row, fetch
// its stock UOM into custom_uom (read-only). Gated on the parent being a
// Stock Entry so the shared Landed Cost Voucher form is untouched.
function _fetch_addl_cost_service_item_uom(frm, cdt, cdn) {
    if (frm.doctype !== "Stock Entry") return;
    const row = locals[cdt][cdn];
    if (!row.custom_service_item) return;
    frappe.db.get_value("Item", row.custom_service_item, "stock_uom", (r) => {
        if (r && r.stock_uom) {
            frappe.model.set_value(cdt, cdn, "custom_uom", r.stock_uom);
        }
    });
}

// CR-07.4 — when a service PO line is picked, fetch its rate into
// custom_rate (mirrors FR-11 on source rows), plus item + UOM, then
// recompute amount through CL-18. The user may still override the rate.
function _fetch_addl_cost_po_item(frm, cdt, cdn) {
    if (frm.doctype !== "Stock Entry") return;
    const row = locals[cdt][cdn];
    if (!row.custom_purchase_order_item) return;
    frappe.db.get_value(
        "Purchase Order Item",
        row.custom_purchase_order_item,
        ["rate", "item_code", "uom"],
        (r) => {
            if (!r) return;
            if (r.item_code && !row.custom_service_item) {
                frappe.model.set_value(cdt, cdn, "custom_service_item", r.item_code);
            }
            if (r.uom) {
                frappe.model.set_value(cdt, cdn, "custom_uom", r.uom);
            }
            if (r.rate) {
                frappe.model.set_value(cdt, cdn, "custom_rate", flt(r.rate));
            }
            // set_value on custom_rate above fires the custom_rate handler
            // which recomputes amount; call again defensively in case rate
            // was unchanged but qty was already present.
            _recompute_addl_cost_amount(frm, cdt, cdn);
        },
    );
}

frappe.ui.form.on("Landed Cost Taxes and Charges", {
    custom_qty: _recompute_addl_cost_amount,
    custom_rate: _recompute_addl_cost_amount,
    custom_service_item: _fetch_addl_cost_service_item_uom,
    custom_purchase_order_item: _fetch_addl_cost_po_item,
});

File: detox_project/detox_project/overrides/stock_entry_manufacture.py
# ABP2-I419 Phase 3 (Sahil 2026-06-17, Out of BRD) — Stock Entry
# Manufacture customizations. Implements the FRD's Process Logic L08..L12
# (process auto-fetch, target defaults, expense/PO rate fetch) and the
# server side of Validations VAL-08..14.
#
# Wired in hooks.py:
#   Stock Entry:
#     validate -> validate_stock_entry_manufacture
# (the CC/Project guard from Phase 2 stays wired separately).

from __future__ import annotations

import frappe
from frappe import _
from frappe.utils import flt


# Stock Entry types covered by Phase 3 logic.
MFG_TYPES = {
    "Manufacture",
    "Material Transfer for Manufacture",
    "Repack",
}


# --------------------------------------------------------------------------
# Whitelisted helper for the Client Script's auto-fetch.
# --------------------------------------------------------------------------
@frappe.whitelist()
def get_operation_rm_rows(production_plan: str, operation: str) -> list[dict]:
    """L08 — return RM/Service rows on the plan for the given operation.

    Used by the SE Manufacture Client Script: when the user picks a
    Process on the SE header, the client clears `items` then calls this
    method to repopulate with the operation's source rows (item, UOM,
    rate, qty_per_unit) — qty starts blank for the user to enter.
    """
    if not production_plan or not operation:
        return []
    rows = frappe.db.sql(
        """SELECT op.item_code, op.item_type, op.standard_uom, op.manual_uom,
                  op.standard_rate, op.qty_per_unit, op.multiply_by,
                  op.cost_center, op.project
           FROM `tabDetox Production Plan Operation` op
           WHERE op.parent = %s
             AND op.parenttype = 'Production Plan'
             AND op.operation_name = %s
           ORDER BY op.idx""",
        (production_plan, operation), as_dict=True,
    )
    enriched = []
    for r in rows:
        uom = r.manual_uom or r.standard_uom
        # Default expense account from Item master (L11).
        expense_account = _default_expense_account(r.item_code)
        # Item name for display.
        item_name = frappe.db.get_value("Item", r.item_code, "item_name") or r.item_code
        enriched.append({
            "item_code": r.item_code,
            "item_name": item_name,
            "item_type": r.item_type,
            "uom": uom,
            "stock_uom": r.standard_uom,
            "basic_rate": flt(r.standard_rate),
            "qty_per_unit": flt(r.qty_per_unit),
            "multiply_by": flt(r.multiply_by),
            "expense_account": expense_account,
            "cost_center": r.cost_center,
            "project": r.project,
        })
    return enriched


def _default_expense_account(item_code: str) -> str | None:
    """Pull the default expense account from Item master (L11)."""
    if not item_code:
        return None
    row = frappe.db.sql(
        """SELECT expense_account FROM `tabItem Default`
           WHERE parent = %s
             AND parenttype = 'Item'
             AND IFNULL(expense_account, '') != ''
           LIMIT 1""",
        item_code,
    )
    return row[0][0] if row else None


@frappe.whitelist()
def get_plan_processes(production_plan: str) -> list[dict]:
    """Sahil Image #30 — the Process dropdown on the SE was empty
    because client-side frappe.db.get_list on the child DocType
    'Detox Production Plan Process' hit permission walls (the user
    has no direct read permission on the child; child perms inherit
    from parent only when accessed via the parent's bag). A
    server-side whitelisted SQL bypasses the issue and returns the
    plan's declared Operations + their Workstation."""
    if not production_plan:
        return []
    return frappe.db.sql(
        """SELECT operation_name, workstation, operation_seq
           FROM `tabDetox Production Plan Process`
           WHERE parent = %s AND parenttype = 'Production Plan'
           ORDER BY operation_seq, idx""",
        production_plan, as_dict=True,
    )


@frappe.whitelist()
def get_fg_defaults(production_plan: str, item_code: str) -> dict:
    """L10 — when a target FG row's item_code is set, return the
    matching Table 1 row's fg_warehouse + standard_costing_rate so the
    Client Script can stamp t_warehouse + basic_rate."""
    if not production_plan or not item_code:
        return {}
    row = frappe.db.sql(
        """SELECT fg_warehouse, standard_costing_rate
           FROM `tabDetox Production Plan FG`
           WHERE parent = %s
             AND parenttype = 'Production Plan'
             AND item_code = %s
           LIMIT 1""",
        (production_plan, item_code), as_dict=True,
    )
    if not row:
        return {}
    return {
        "t_warehouse": row[0].fg_warehouse,
        "basic_rate": flt(row[0].standard_costing_rate),
    }


# --------------------------------------------------------------------------
# Inherit dimensions from the linked Production Plan
# --------------------------------------------------------------------------
def inherit_se_from_production_plan(doc, method=None):
    """Sahil 2026-06-17 — when the SE is linked to a Production Plan,
    pull CC + Project off the plan onto the SE header AND propagate to
    every item row that's missing a cost_center. This stops ERPNext's
    get_default_cost_center from ever needing to fall back to
    Company.default_cost_center (the source of the 'Please set default
    Default Cost Center in Company …' throw).

    Fires on validate so it runs BEFORE ERPNext's standard validation.

    Also folds in the hidden `cost_center` proxy field (added so legacy
    Client Scripts can set_value 'cost_center' without erroring) —
    whatever lands there is copied into custom_cost_center.
    """
    if not _stock_entry_is_in_scope(doc):
        return

    # Fold the proxy value first, in case the plan path doesn't fire.
    proxy_cc = doc.get("cost_center")
    if proxy_cc and not doc.get("custom_cost_center"):
        doc.custom_cost_center = proxy_cc

    plan_name = doc.get("production_plan")
    if not plan_name or not frappe.db.exists("Production Plan", plan_name):
        return
    plan = frappe.db.get_value(
        "Production Plan", plan_name,
        ["custom_cost_center", "project"],
        as_dict=True,
    ) or {}
    cc = plan.get("custom_cost_center")
    pj = plan.get("project")
    if cc and not doc.get("custom_cost_center"):
        doc.custom_cost_center = cc
    if pj and not doc.get("project"):
        doc.project = pj
    # Propagate to every item row — blank rows inherit, populated rows
    # are left alone.
    final_cc = doc.get("custom_cost_center") or cc
    final_pj = doc.get("project") or pj
    for row in (doc.get("items") or []):
        if final_cc and not row.get("cost_center"):
            row.cost_center = final_cc
        if final_pj and not row.get("project"):
            row.project = final_pj


# --------------------------------------------------------------------------
# Validate hook
# --------------------------------------------------------------------------
def validate_stock_entry_manufacture(doc, method=None):
    """Phase 3 validations (VAL-08, VAL-10, VAL-11, VAL-12, VAL-13).

    All fire ONLY on Manufacturing-flow SEs. CC + Project enforcement
    stays in cc_project_guard.validate_stock_entry.
    """
    purpose = (doc.get("stock_entry_type") or "").strip()
    if purpose not in MFG_TYPES:
        return

    # VAL-08 — process_selection mandatory before save / submit.
    if doc.get("production_plan") and not doc.get("custom_process_selection"):
        frappe.throw(
            _("Select a Process before adding production rows."),
            title=_("Process Selection missing"),
        )

    # VAL-11 — production_time bounds.
    pt = flt(doc.get("custom_production_time"))
    if pt:
        uom = (doc.get("custom_time_uom") or "Hours").strip()
        cap = 24.0 if uom == "Hours" else 24 * 60.0
        if pt <= 0 or pt >= cap:
            frappe.throw(
                _("Production Time must be greater than 0 and within a "
                  "single day for the selected unit ({0}).").format(uom),
                title=_("Production Time out of range"),
            )

    # VAL-12 — at least one finished/semi-finished target row with qty > 0
    # on a Manufacture entry.
    if purpose == "Manufacture":
        has_fg = any(
            row.get("is_finished_item") and flt(row.get("qty")) > 0
            for row in (doc.get("items") or [])
        )
        if not has_fg:
            frappe.throw(
                _("Enter at least one finished good with quantity to record "
                  "production."),
                title=_("Finished Good missing"),
            )

    # VAL-10 — every source row must carry a PO link on submit. Block
    # Service rows hard, warn on RM (we use ValidationError block for
    # both — adjust if Sahil reopens with a softer policy).
    is_submit_path = (
        doc.docstatus == 1
        or getattr(doc, "_action", None) == "submit"
        or getattr(doc.flags, "validate_before_submit", False)
    )
    if is_submit_path:
        for idx, row in enumerate(doc.get("items") or [], start=1):
            # Source rows have s_warehouse set.
            if not row.get("s_warehouse"):
                continue
            if not row.get("custom_purchase_order") or not row.get("custom_purchase_order_item"):
                frappe.throw(
                    _("Row #{0}: link a Purchase Order and PO line on the "
                      "source row for accurate variance.").format(idx),
                    title=_("Purchase Order missing on source row"),
                )

    # VAL-13 — over-production warning (cumulative produced > planned).
    # Read total_produced from the matching Table 1 row; soft warning only.
    if purpose == "Manufacture" and doc.get("production_plan"):
        _warn_on_over_production(doc)


def _warn_on_over_production(doc) -> None:
    plan = doc.get("production_plan")
    if not plan:
        return
    # Per FG row in the SE, look up Table 1 planned qty + total_produced.
    for row in (doc.get("items") or []):
        if not row.get("is_finished_item"):
            continue
        qty = flt(row.get("qty"))
        if qty <= 0:
            continue
        planned = frappe.db.sql(
            """SELECT qty_to_manufacture, custom_total_produced
               FROM `tabDetox Production Plan FG`
               WHERE parent = %s AND parenttype = 'Production Plan'
                 AND item_code = %s
               LIMIT 1""",
            (plan, row.item_code), as_dict=True,
        )
        if not planned:
            continue
        plan_qty = flt(planned[0].qty_to_manufacture)
        produced = flt(planned[0].custom_total_produced)
        if (produced + qty) > plan_qty:
            frappe.msgprint(
                _("This entry exceeds the planned quantity for {0} "
                  "({1}/{2}).").format(
                    row.item_code, produced + qty, plan_qty,
                ),
                title=_("Over-production"),
                indicator="orange",
            )


# --------------------------------------------------------------------------
# total_produced rollup — L07 (kept in Phase 3 alongside the related
# validations; called from Stock Entry on_submit / on_cancel).
# --------------------------------------------------------------------------
def rollup_total_produced_on_submit(doc, method=None):
    purpose = (doc.get("stock_entry_type") or "").strip()
    if purpose not in MFG_TYPES or not doc.get("production_plan"):
        return
    _adjust_total_produced(doc, sign=+1)


def rollup_total_produced_on_cancel(doc, method=None):
    purpose = (doc.get("stock_entry_type") or "").strip()
    if purpose not in MFG_TYPES or not doc.get("production_plan"):
        return
    _adjust_total_produced(doc, sign=-1)


def _adjust_total_produced(doc, sign: int) -> None:
    plan = doc.get("production_plan")
    for row in (doc.get("items") or []):
        if not row.get("is_finished_item"):
            continue
        qty = flt(row.get("qty"))
        if qty <= 0:
            continue
        # Update Table 1 row matching this FG item.
        frappe.db.sql(
            """UPDATE `tabDetox Production Plan FG`
               SET custom_total_produced = COALESCE(custom_total_produced, 0) + %s
               WHERE parent = %s AND parenttype = 'Production Plan'
                 AND item_code = %s""",
            (sign * qty, plan, row.item_code),
        )
